### DISCUSSION TITLE
The GOS differences between GOS = ebsd.grainMean(mis2mean.angle, grains) and grains.GOS./ degree

### DISCUSSION THREAD

**Discussion by PengJin-SRAS** [mtex-toolbox/mtex#2226] [Ask Anything]
Hi Community,

I am trying to plot GOS in MTEX(mtex-6.0.beta3), but I found the plots are different on the GOS values on certain grains, and this happens at my other samples' plots as well:

I started with grain reconstruction:

```
% compute grains-------------------global grain calculation-------------------
ebsd_region1 = ebsd(inpolygon(ebsd,region1));
[grains,ebsd_region1.grainId] = calcGrains(ebsd_region1('indexed'),'angle',10*degree);
% remove small grains
ebsd_region1(grains(grains.grainSize < 3)) = [];
% and repeat the grain computation
[grains,ebsd_region1.grainId] = calcGrains(ebsd_region1('indexed'),'angle',10*degree);
grains = smooth(grains,5);
plot(ebsd_region1('Fe-FCC'),ebsd_region1('Fe-FCC').orientations);

```

Then, I used the variables `grains.GOS` after grain construction and plot via `plot(grains, grains.GOS./ degree)`:

![image](https://github.com/user-attachments/assets/2dc72d1c-53fa-4fa1-ac43-9299279938a6)

When I tried to recalculate the GOS via 

```
mis2mean = calcGROD(ebsd_region1('indexed'), grains('indexed')); % computes the misorientations to the grain mean orientation
GOS = ebsd_region1('indexed').grainMean(mis2mean.angle, grains('indexed'));
plot(grains, GOS ./ degree)
```
Then I got:

![image](https://github.com/user-attachments/assets/5c666978-399e-4d36-a753-06259ee2b9a5)

You can see, a different GOS value on the left big grains region, and the maximum value on the colourbar also varies. I am wondering why that happens and whether this relates to the MTEX version. Which one is convinced?

Thank you for your time.

Kind regards,

Peng

**Comment by anamitrasikdar** [comment]
Hi Peng,

I am not sure but I guess that the problem is because that grain is poorly indexed if compared to the other grains. Could you please write here what was the indexing rate of the grain that you find behaving oddly?

I mean,
iRate = number of indexed ebsd pixels/grainArea

Best
Anamitra

**Reply by zmichels** [reply]
It might be that the grains.GOS uses the @max, rather than the @mean method(?). If you try ebsd.grainMean(mis2mean.angle, grains, @max)... are the outputs more similar?

**Reply by anamitrasikdar** [reply]
Hi Zach,

This is the part where GOS is calculated. Am I right?

% compute mean orientation and GOS
if 0
  GOS = zeros(length(grains),1); %#ok<UNRCH>
  doMeanCalc = find(grains.grainSize>1 & grains.isIndexed);
  abcd = zeros(length(doMeanCalc),4);
  for k = 1:numel(doMeanCalc)
    qind = subSet(q,d(grainRange(doMeanCalc(k))+1:grainRange(doMeanCalc(k)+1)));
    mq = mean(qind,'robust');
    abcd(k,:) = [mq.a mq.b mq.c mq.d];
    GOS(doMeanCalc(k)) = mean(angle(mq,qind)); 
  end
  meanRotation(doMeanCalc)=reshape(quaternion(abcd'),[],1);
else
  %[meanRotation, GOS] = accumarray(grainId(grainId>0),q(grainId>0),'robust');
  [meanRotation, GOS] = accumarray(grainId(grainId>0),q(grainId>0));
end
% save 
grains.prop.GOS = GOS;



I am wondering where it is taking the max method by default? And I also want to know why there is an if 0 switch to not execute that part...


Best,
Anamitra

**Reply by PengJin-SRAS** [reply]
Hi Anamitra and Zach,

Here are some results based on your suggestions:

@anamitrasikdar:

I plot two indexing rate of the grains which bahaved oddly:

1. Grain 1 

```
grainId = 2692
phase = Fe-FCC
(x,y) = 363, 1102, 1
Euler = (76.2,96.6,132)
``` 
![image](https://github.com/user-attachments/assets/f143d451-9fd0-4f38-bdfd-2b30131cec85)

```
rate1=grains(2692).grainSize/grains(2692).area
rate1 =

    0.0290
```

```
grains(2692).GOS./degree
ans =

   30.7459

GOS(2692)./degree
ans =

   24.9278
```


2. Grain 2

```
grainId = 2385
phase = Fe-FCC
(x,y) = 204, 1081, 1
Euler = (7,55.6,251.8)
```
![image](https://github.com/user-attachments/assets/88389095-7347-4543-83e2-3dd827bf1971)
 
```
rate2=grains(2385).grainSize/grains(2385).area`

rate2 =

    0.0299

```

```
grains(2385).GOS./degree
ans =

   27.5142

GOS(2385)./degree
ans =

   24.2586
```
3. And I do the same thing for a grain which behaved relatively normally： 
Grain 3:

```
grainId = 2694
phase = Fe-FCC
(x,y) = 1216, 1489, 1
Euler = (337.8,92.3,358.8)

```
 
![image](https://github.com/user-attachments/assets/97ec982f-641e-4667-9bb1-8e2f6f8bf02f)

```
rate3=grains(2694).grainSize/grains(2385).area
rate3 =

    0.0869

grains(2694).GOS./degree
ans =

   30.5554

GOS(2694)./degree
ans =

   30.4570
```

Grain 1 and 2 has relatively low indexing rate than grain 3. But the value 0.02 is reasonable or not? If not, how can I improve this? And even if a higher indexing rate of 0.08, there still is a differences on GOS based on two calculation method.

Hi, @zmichels 
I think you are talking about the maximum of GOS, which was plot in the figure below as MGOS:

![image](https://github.com/user-attachments/assets/5f463e96-a603-4cc5-bd54-81d20ab6a4fc)

```
MGOS = ebsd_region1('indexed').grainMean(mis2mean.angle, grains, @max);
plot(grains, MGOS ./ degree)
```
Any suggestions?

Thank you for your time!

Kind regards,

Peng

**Reply by anamitrasikdar** [reply]
Hi Peng,

I think what you are trying to calculate is not the indexing rate. It giving you inverse of the area of 1 pixel.

Please refer to the supplementary of Cross et al, 2017.

**Reply by zmichels** [reply]
I unclear how the issue would be related to indexing rate (but I'm likely missing something) when both values are being calculated from the same dataset (not changing the indexing rate between computing one versus the other). The question seems to be more about how to replicate the GOS value... and the proposed approach isn't working (so it must not match, no?).

**Reply by zmichels** [reply]
I'll note while trying some things that the following results are different -- something is dissimilar in the way GROD is computed versus "mis2mean" when grains are computed... which is when grains.GOS is computed and assigned -- so perhaps when you are using "GROD" as "mis2mean" above, it will yield a different result than if using the mis2mean rotations computed during grain reconstruction:

```matlab
mtexdata forsterite

% compute grains
[grains,ebsd.grainId,ebsd.mis2mean]  = calcGrains(ebsd,'threshold',[1 10]*degree)


% mis2mean from calcGrains
mis2mean = ebsd.mis2mean;

% compute GROD
grod = ebsd.calcGROD(grains);

% compare GROD angle with mis2mean angle
figure,
histogram([mis2mean.angle./degree])
hold on
histogram([grod.angle./degree])


%% "GOS" comparisons
% GOS from calcGrains
GOS = grains.GOS;

% compute the mean m2m angles for each grain
m2mGOS = ebsd.grainMean(mis2mean.angle, grains, @mean);

% compute the mean GROD angles for each grain
grodGOS = ebsd.grainMean(grod.angle, grains, @mean);


%% plots?

figure,
plot(grains, GOS./ degree)

nextAxis
plot(grains, m2mGOS./degree)

nextAxis
plot(grains, grodGOS./degree)
```

**Reply by PengJin-AM** [reply]
Hi all,

Many thanks for your reply. 

‘I'll note while trying some things that the following results are different ’- So, it is likely to be something unexpected happened in the codes?

Any suggestions please? how do I know which calculation is right. As it will affect the results of maximum misorientation angles (MGOS) and Grain average misorientation (GAM), which are calculated from 'calcGROD'.

Thank you.

Kind regards,

Peng
